client/engine/chainservice/mock_chain.go

```diff
@@ -0,0 +1,74 @@
+package chainservice
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/statechannels/go-nitro/client/engine/store/safesync"
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/types"
+)
+
+// MockChain is an in-memory "chain" that accepts transactions and broadcasts events
+type MockChain interface {
+	SubmitTransaction(protocols.ChainTransaction) error // unlike an ethereum blockchain, Mockhain accepts go-nitro protocols.ChainTransaction
+	SubscribeToEvents(a types.Address) <-chan Event     // returns a channel that produces all chain Events
+}
+
+// MockChainImpl mimicks the Ethereum blockchain by keeping track of block numbers and account balances.
+type MockChainImpl struct {
+	blockNum uint64
+	// holdings tracks funds for each channel.
+	holdings map[types.Destination]types.Funds
+	// out maps addresses to an Event channel. Given that MockChainServices only subscribe
+	// (and never unsubscribe) to events, this can be converted to a list.
+	out safesync.Map[chan Event]
+}
+
+// NewMockChainImpl creates a new MockChainImpl
+func NewMockChainImpl() *MockChainImpl {
+	chain := MockChainImpl{}
+	chain.blockNum = 1
+	chain.holdings = make(map[types.Destination]types.Funds)
+	chain.out = safesync.Map[chan Event]{}
+	return &chain
+}
+
+// SubmitTransaction updates internal state and brodcasts events
+func (mc *MockChainImpl) SubmitTransaction(tx protocols.ChainTransaction) error {
+	mc.blockNum++
+	switch tx := tx.(type) {
+	case protocols.DepositTransaction:
+		if tx.Deposit.IsNonZero() {
+			mc.holdings[tx.ChannelId()] = mc.holdings[tx.ChannelId()].Add(tx.Deposit)
+		}
+		for address, amount := range tx.Deposit {
```

Comment:
Is `address` a token address or participant address? I think it's a token address -- renaming it to `assetAddress` might be helpful.

In either case, it's surprising to me that one deposit could generate multiple events. Is this consistent with `nitro-protocol`?

Reply:
`address` is renamed to `assetAddress` in d744a989.

My understanding is: for a channel that contains `n` token types, `n` deposit transactions are submitted to the blockchain. This will result in `n` chain events. This code mimics that behavior. Do you think I am misinterpreting nitro-protocol?

Reply:
ah yes, you're totally correct!